New mode for compiling SQL queries from DDlog's query syntax

netj opened the pull request:
`ddlog query` for implementing `deepdive query` in HazyResearch/deepdive#476

netj committed:
Adds sql command to generate SQL from queries in `?- foo(x,y,...).` like syntax

netj committed:
Renames to ddlog query, supports only a single conjunctive query body

no disjunctions separated with ";" for the query.

netj committed:
Runs SemanticChecker on queries as well with pretty printing `?-`

netj committed:
Adds a check to prevent extra rules in queries redefining existing relation in the program

netj committed:
Adds a test for ddlog query

netj committed:
Makes the generated SQL for a query use a sensible column name

than just column_0 and _1, etc.

netj committed:
Fixes failing test's actual equivalent output as expected

netj committed:
Fixes `make actual-expected` for accepting actual outputs as expected

which can be handy to see the difference with `git diff` afterwards.

netj committed:
Adds more tests for ddlog query

netj committed:
Supports @order_by on head terms for ddlog query

`@order_by`, `@order_by("ASC")`, `@order_by(dir="ASC")` all work.
`@order_by(priority=0)` or `@order_by("ASC", 0)` work to specify the key
index/priority when more than one head terms are used for ordering.

TODO make normal rules also support @order_by

netj committed:
Fixes some whitespace/ordering changes in tests' expected outputs

No idea why some Scala Lists change order (maybe they're Sets), maybe we
should sort them to ensure a normal order.

feiranwang commented:
The `order_by` annotation takes a different form from other annotations. Looks good other than this.

netj commented:
Not sure what you mean by different form?  That it parses with the general annotation parser also used for schema annotations?  I think the rest of annotations in the future should be parsed in that way, maybe except `@weight` and `@label` that take general expressions.

feiranwang commented:
By different form I mean there is a `@order_by num_candidates`, and I was assuming something like `@order_by(num_candidates)`. Maybe the former is also allowed in the general annotations?

netj commented:
Those two are separate: an annotation (`@order_by`) and an expression (`num_candidates`).  I extended the grammar (at least for the query rules) to accept optional annotations before each head term expression.  See my examples in HazyResearch/deepdive#476.  So `@foo @order_by @bar(123,456) num_candidates, @bar(78,90) another_var ?- ...` is also valid.

Query rules are special case and the first won't get confused with rule annotations.  For normal rules (in the future when we extend the grammar), I think there won't be ambiguity because the head name and a parenthesis will also be there before each head term expression, right?

There's a limitation that you can't hide the order by column of course but I think people will like to just prefix a column with an annotation rather than repeating a variable or expression to simply order by.  If we think people will frequently want to drop the ordering column, we should go ahead and support `@order_by` for normal rules as well, so a temporary table/view can be defined first then projected out in the actual query.

feiranwang commented:
I see. Thanks for the explanation!

> On Feb 1, 2016, at 2:04 AM, Jaeho Shin notifications@github.com wrote:
> 
> Those two are separate: an annotation (@order_by) and an expression (num_candidates). I extended the grammar (at least for the query rules) to accept optional annotations before each head term expression. See my examples in HazyResearch/deepdive#476. So @foo @order_by @bar(123,456) num_candidates, @bar(78,90) another_var ?- ... is also valid.
> 
> Query rules are special case and the first won't get confused with rule annotations. For normal rules (in the future when we extend the grammar), I think there won't be ambiguity because the head name and a parenthesis will also be there before each head term expression, right?
> 
> There's a limitation that you can't hide the order by column of course but I think people will like to just prefix a column with an annotation rather than repeating a variable or expression to simply order by. If we think people will frequently want to drop the ordering column, we should go ahead and support @order_by for normal rules as well, so a temporary table/view can be defined first then projected out in the actual query.
> 
> —
> Reply to this email directly or view it on GitHub.

netj commented:
(I'm resolving the merge conflicts now btw)

netj committed:
Merge branch 'master' of https://github.com/HazyResearch/ddlog into ddlog-query

netj commented:
Oops the lower coverage shows some dead code and sloppy merge of mine.  Will push another commit.  Please merge later!

netj committed:
Eliminates some premature merge that left dead code

And supports multiple disjunctive bodies in query, taking the common
variables among all bodies to project, or all columns with `*`.

netj committed:
Adds tests for multiple bodies in query and fixes a parser bug

that tried to parse conjunctive query atom as function call expression :S

netj committed:
Fixes changed expected outputs, adds a few more test

netj commented:
Okay ready to merge.  Coverage drop seems negligible.

bar commented:
@netj WAT?
